src/token.rs

```diff
@@ -662,80 +886,174 @@ impl Token {
+                pParameter: &mut params as *mut interface::CK_GCM_PARAMS
+                    as *mut _,
+                ulParameterLen: std::mem::size_of::<CK_GCM_PARAMS>()
+                    as CK_ULONG,
+            };
+            let aes = self.mechanisms.get(CKM_AES_GCM)?;
+            let mut op = aes.encryption_new(&mech, &kek)?;
+            let mut clen: CK_ULONG =
+                op.encryption_len(val.len() as CK_ULONG)? as CK_ULONG + 12;
+            let mut cipher = Vec::<u8>::with_capacity(clen as usize);
+            cipher.extend_from_slice(&iv);
+            cipher.resize(clen as usize, 0);
+            clen -= iv.len() as CK_ULONG;
+            op.encrypt(
+                val.as_slice(),
+                cipher.as_mut_slice()[12..].as_mut_ptr(),
+                &mut clen,
+            )?;
+
+            unsafe { cipher.set_len(clen as usize + 12) };
+            return Ok(cipher);
+        } else {
+            return err_rv!(CKR_GENERAL_ERROR);
+        }
+    }
 
-        CKR_OK
+    fn object_to_storage(
+        &mut self,
+        mut obj: Object,
+        encrypt: bool,
+    ) -> KResult<()> {
+        let uid = obj.get_attr_as_string(CKA_UNIQUE_ID)?;
+        if encrypt {
+            let ats = self.object_factories.get_sensitive_attrs(&obj)?;
+            for typ in ats {
+                let plain = obj.get_attr_as_bytes(typ)?;
+                let encval = self.encrypt_value(&uid, plain)?;
+
+                /* now replace the clear text val with the encrypted one */
+                obj.set_attr(attribute::from_bytes(typ, encval))?;
+            }
+        }
+        self.storage.store(&uid, obj)
     }
 
-    pub fn save(&mut self) -> KResult<()> {
-        self.storage.flush()
+    fn decrypt_value(&self, uid: &String, val: &Vec<u8>) -> KResult<Vec<u8>> {
+        if let Some(ref kek) = self.kek {
+            let aad = uid.as_bytes();
+            let tag_len = 8;
+            let mut params = CK_GCM_PARAMS {
+                pIv: val.as_ptr() as *mut CK_BYTE,
+                ulIvLen: 12,
+                ulIvBits: 96,
+                pAAD: aad.as_ptr() as *mut CK_BYTE,
+                ulAADLen: aad.len() as CK_ULONG,
+                ulTagBits: (tag_len * 8) as CK_ULONG,
+            };
+            let mech: CK_MECHANISM = CK_MECHANISM {
+                mechanism: CKM_AES_GCM,
+                pParameter: &mut params as *mut interface::CK_GCM_PARAMS
+                    as *mut _,
+                ulParameterLen: std::mem::size_of::<CK_GCM_PARAMS>()
+                    as CK_ULONG,
+            };
+            let aes = self.mechanisms.get(CKM_AES_GCM)?;
+            let mut op = aes.decryption_new(&mech, &kek)?;
+            let mut plen: CK_ULONG =
+                op.decryption_len((val.len() - 12) as CK_ULONG)? as CK_ULONG;
+            let mut plain = Vec::<u8>::with_capacity(plen as usize);
+            op.decrypt(&val.as_slice()[12..], plain.as_mut_ptr(), &mut plen)?;
```

Comment:
same as the encryption -- the magic number 12 should be at least variable, better some constant.